### PR TITLE
Convert case internally : Dockerfile to dockerfile

### PR DESCRIPTION
**Pull request by itscaro**
## Motivation and Context

- [x] I have raised an issue to propose this change (#317)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by rgee0** [comment]
I'm really not sure about the merits of deprecating `Dockerfile`.  The whole world is already far too familiar with a capital D.  

It does feel like we should be able to be agnostic to case without too much of a fuss.

**Comment by alexellis** [comment]
@itscaro we need to update this - https://github.com/openfaas/faas-cli/blob/master/samples.yml#L36 and this - https://github.com/openfaas/faas/blob/master/sample-functions/samples.yml

File: commands/build.go [hunk 1]
```diff
@@ -65,7 +65,19 @@ via flags.`,
...
-	RunE: runBuild,
+	PreRunE: preRunBuild,
+	RunE:    runBuild,
+}
+
+// preRunBuild validates args & flags
+func preRunBuild(cmd *cobra.Command, args []string) error {
+	if language == "Dockerfile" {
+		// TODO DEPRECATION: special case to handle --lang Dockerfile
+		fmt.Println(aec.YellowF.Apply("/!\\ DEPRECATION NOTICE: Please use 'dockerfile' as language instead of 'Dockerfile', language will accept only lowercases in future releases"))
```
Comment (alexellis) [inline commands/build.go]: Can we avoid duplicating the message 3x times?

Comment (alexellis) [reply commands/build.go]: What does `/!\\` do? Could Deprecation notice be less shouty/loud?

File: commands/deploy.go [hunk 2]
```diff
@@ -89,7 +90,19 @@ via flags. Note: --replace and --update are mutually exclusive.`,
...
                   --gateway=http://remote-site.com:8080 --lang=python
                   --env=MYVAR=myval`,
-	RunE: runDeploy,
+	PreRunE: preRunDeploy,
+	RunE:    runDeploy,
+}
+
+// preRunDeploy validates args & flags
+func preRunDeploy(cmd *cobra.Command, args []string) error {
+	if language == "Dockerfile" {
```
Comment (alexellis) [inline commands/deploy.go]: What happens if someone is using a YAML file rather than CLI args?

**Comment by itscaro** [comment]
@alexellis I forgot to re-integrate the check in build.go, just did it and removed also the special case in language_template.go `IsValidTemplate()`

File: commands/common.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+package commands
+
+import (
+	"fmt"
+
+	"github.com/morikuni/aec"
+)
+
+func validateLanguageFlag(language *string) {
```
Comment (itscaro) [inline commands/common.go]: move to common.go to remove repetition

File: commands/new_function.go [hunk 4]
```diff
@@ -18,12 +18,11 @@ import (
 
 var (
 	appendFile string
-	lang       string
 	list       bool
 )
 
 func init() {
-	newFunctionCmd.Flags().StringVar(&lang, "lang", "", "Language or template to use")
+	newFunctionCmd.Flags().StringVar(&language, "lang", "", "Language or template to use")
```
Comment (itscaro) [inline commands/new_function.go]: use the global var `language` instead

**Comment by itscaro** [comment]
@rgee0 docker image is lower case, for language like PHP which is mostly referred in upper case, in the docker world, it's `php` instead
We can apply this rule to language template too.

```
docker pull PHP
invalid reference format: repository name must be lowercase
```

**Comment by rgee0** [comment]
@itscaro I am thinking more about convention.  Looking at all our repos, we follow the convention, uppercase D for `Dockerfile`.  From my perspective if I'm used to typing `Dockerfile` and within OF I need to remember its `dockerfile` then I would find that hugely frustrating.

**Comment by itscaro** [comment]
@alexellis @rgee0 So I will make change to YAML files: dockerfile to Dockerfile, is that correct before I do the change?

**Comment by rgee0** [comment]
See also #320

**Comment by alexellis** [comment]
@itscaro I updated those YAML files.. do you know if there are any others which need doing?

**Comment by alexellis** [comment]
I also merged #320 as it seemed like a smaller change to put into a minor patch release.

**Comment by itscaro** [comment]
@alexellis it covers only new_function case instead of the globality of this bug.

File: commands/common.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+	"errors"
+)
+
+func validateLanguageFlag(language string) (string, error) {
+	var err error = nil
+
+	if language == "Dockerfile" {
+		language = "dockerfile"
+
+		err = errors.New("Dockerfile was converted to dockerfile")
```
Comment (alexellis) [inline commands/common.go]: What is this error for? ( I thought we were re-moving the message? )

File: commands/build.go [hunk 6]
```diff
@@ -115,7 +123,7 @@ func build(services *stack.Services, queueDepth int, shrinkwrap bool) {
 			for function := range workChannel {
 				fmt.Printf(aec.YellowF.Apply("[%d] > Building %s.\n"), index, function.Name)
 				if len(function.Language) == 0 {
-					fmt.Println("Please provide a valid --lang or 'Dockerfile' for your function.")
+					fmt.Println("Please provide a valid language for your function.")
```
Comment (alexellis) [inline commands/build.go]: looks fine

**Comment by alexellis** [comment]
Makes sense @itscaro .. @ericstoekl Quan is saying that we may have missed one or two cases in the previously merged PR. PTAL

File: commands/common.go (hunk 5 above)
Comment (itscaro) [reply commands/common.go]: Just in case we need to display the event, it's not a real error, just a notice to know that we modified the language variable

**Comment by alexellis** [comment]
Hi @itscaro I'm looking to merge this now

**Pull request merged**